### PR TITLE
Add TimmBackbone model

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Adds a new model TimmBackbone to use for loading timm weights for use in the AutoBackbone API.

Example usage: 
```
from transformers import AutoBackbone

# Loads a transformers model 
backbone = AutoBackbone.from_pretrained("microsoft/resnet-18")

# Loads a timm checkpoint
backbone = AutoBackbone.from_pretrained("resnet18", use_timm_backbone=True)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: .circleci/create_circleci_config.py [hunk 1]
```diff
@@ -445,6 +445,7 @@ def job_name(self):
         "pip install -e .[dev]",
         "pip install git+https://github.com/huggingface/accelerate",
         "pip install --upgrade pytest pytest-sugar",
+        "pip install natten",
```
Comment (amyeroberts) [inline .circleci/create_circleci_config.py:448]: Needed for Dinat and Nat doc tests

File: src/transformers/models/auto/auto_factory.py [hunk 2]
```diff
@@ -492,6 +492,45 @@ def register(cls, config_class, model_class):
...
+    @classmethod
+    def _load_timm_backbone_from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
+        requires_backends(cls, ["vision", "timm"])
+        from ...models.timm_backbone import TimmBackboneConfig
+
+        config = kwargs.pop("config", TimmBackboneConfig())
+
+        use_timm = kwargs.pop("use_timm_backbone", True)
+        if not use_timm:
+            raise ValueError("use_timm_backbone must be True for timm backbones")
```
Comment (sgugger) [inline src/transformers/models/auto/auto_factory.py:508]:
```suggestion
            raise ValueError("`use_timm_backbone` must be `True` for timm backbones")
```

File: src/transformers/models/auto/auto_factory.py [hunk 3]
```diff
@@ -492,6 +492,45 @@ def register(cls, config_class, model_class):
...
+
+        if kwargs.get("out_features", None) is not None:
+            raise ValueError("Cannot specify out_features for timm backbones")
```
Comment (sgugger) [inline src/transformers/models/auto/auto_factory.py:511]:
```suggestion
            raise ValueError("Cannot specify `out_features` for timm backbones")
```

File: src/transformers/models/timm_backbone/modeling_timm_backbone.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    other models in the library keeping the same API.
+    """
+
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = False
+    config_class = TimmBackboneConfig
+
+    def __init__(self, config, **kwargs):
+        requires_backends(self, "timm")
+        import timm
```
Comment (sgugger) [inline src/transformers/models/timm_backbone/modeling_timm_backbone.py:41]: Might be cleaner to do this at the top of the file under an `is_timm_available()` but no strong opinions.

**Review by sgugger — APPROVED** [review]
Thanks a lot for your PR! Just a couple of nits here and there but it looks great!

**Comment by amyeroberts** [comment]
@ydshieh Merging now. I'll address any comments or requests for changes in a follow up PR.

**Pull request merged**